### ISSUE TITLE
Errors while using "random_points_on_mesh.h"

### ISSUE THREAD

**Issue by Salingo** [libigl/libigl#842]
Hello,
I met a strange issue while using function _random_points_on_mesh_.
I am trying to load different meshes and sample them into points one by one, but the program will crash after I repeat this operation several times. My code looks like:
```
std::vector<Eigen::MatrixXd> V;
std::vector<Eigen::MatrixXi> F;
if (ImGui::Button("Load object"))
{
  V.swap(std::vector<Eigen::MatrixXd>());
  F.swap(std::vector<Eigen::MatrixXi>());
  ... // Load new mesh
}
if (ImGui::Button("Sample & Save Points"))
{
  Eigen::VectorXi FI;
  Eigen::SparseMatrix<double> B;
  igl::random_points_on_mesh(8000, V[0], F[0], B, FI);
}
```
When I use debug mode, the issue occurs at line 69 in _random_points_on_mesh.h_:
![1](https://user-images.githubusercontent.com/15212673/43036168-993b9342-8d2e-11e8-9b37-d20e5b80646f.png)
When I use release mode, the issue occurs at line 1036 in _SparseMatrix.h_ (Eigen):
![2](https://user-images.githubusercontent.com/15212673/43036176-dfb2ce76-8d2e-11e8-9da3-795cd4b9a494.png)

But if I restart the program and sample the same object again, normally it will be ok.
So I'm so confused about this. Hope you can help me, thanks very much.

**Comment by jdumas** [comment]
Could it be you're calling the method on a mesh with empty faces `F`? Looks like it could be problem in how you deal with multiple meshes.

**Comment by Salingo** [comment]
@jdumas Thanks for your reply, I use the following code to make sure the problem is not in V or F:
```
...//read a mesh
count = 0;
while(++count)
{
  igl::random_points_on_mesh(8000, V[1], F[1], B, FI);
}
```
The program will crash when count=4, I don't know why, maybe this is a bug?

**Comment by jdumas** [comment]
Does it crash at the same place? Could you send a MWE along with the meshes you are testing with?

**Comment by Salingo** [comment]
@jdumas Yes, it crash at the same place: line 69 in _random_points_on_mesh.h_, here are my obj file and the simple test code, thanks!

```
Eigen::MatrixXd V;
Eigen::MatrixXi F;
Eigen::VectorXi FI;
Eigen::SparseMatrix<double> B;
igl::readOBJ(".\Cutter_7_0.obj", V, F);
int count = 0;
while(++count)
{
  igl::random_points_on_mesh(8000, V, F, B, FI);
}
```
[obj.zip](https://github.com/libigl/libigl/files/2217689/obj.zip)

**Comment by jdumas** [comment]
I am able to run this code without any issue. I compiled in `DEBUG` mode with gcc 8 on archlinux, and it runs without issues more than 200 times, and also goes through `valgrind` withtout any trouble.

**Comment by Salingo** [comment]
@jdumas Thanks for your tests, I'm using Visual Studio 2017 on Windows 10 and the issue still exists. I built the project using libigl-example-project, here is the MWE, maybe you can test the code directly.
[libigl-example-project.zip](https://github.com/libigl/libigl/files/2226146/libigl-example-project.zip)
(The path of libigl and my project is _D:\Code\Github_)

**Comment by Salingo** [comment]
Hello, I repacked a MWE which does not need additional configration for you to verify, just open the sln with VS2017 and compile the project, then the crash may appear while running the debugger.
Hope you can solve this confusing issue. Thanks!

[MWE.zip](https://github.com/libigl/libigl/files/2504961/MWE.zip)

**Comment by jdumas** [comment]
Thanks! I haven't had time to try it under Windows yet. Can't tell you when that will be the case sorry =)

**Comment by baymin182** [comment]
can I use the function random_points_on_mesh.h to convert different meshes into meshes whose count of vertices are same?

**Comment by jdumas** [comment]
Well, random_points_on_mesh gives you a point cloud, not the connectivity, so it's not a straightforward application.

**Comment by rozentill** [comment]
@Salingo @jdumas  Same problem, crashed after several times in Visual Studio 17 Release mode, did you figure it out?

**Comment by Salingo** [comment]
@rozentill Unfortunately, I haven't figured it out yet.

**Comment by alecjacobson** [comment]
Likely fix: https://github.com/libigl/libigl/pull/1345
